### PR TITLE
Optimize indexer

### PR DESCRIPTION
**Pull request by wojtek-t**
Before the change:
```
BenchmarkIndexer-12        447849        2738 ns/op       242 B/op        4 allocs/op
PASS
ok   k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache 2.451s
```

After the change:
```
BenchmarkIndexer-12    	  906784	 1274 ns/op	  50 B/op	  3 allocs/op
PASS
ok   k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache	2.148s
```

```release-note
NONE
```

/kind cleanup
/priority important-soon
/sig api-machinery

/assign @jpbetz @caesarxuchao

### CODE REVIEW

**Review by jpbetz — COMMENTED** [review]
One very minor nit then LGTM.  Thanks for improving the code organization and adding the benchmark.

File: staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go [hunk 1]
```diff
@@ -254,58 +254,70 @@ func (c *threadSafeMap) AddIndexers(newIndexers Indexers) error {
 // updateIndices modifies the objects location in the managed indexes, if this is an update, you must provide an oldObj
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go:254]: done

**Comment by wojtek-t** [comment]
@jpbetz - thanks! PTAL

**Comment by jberkus** [comment]
@wojtek-t if this is really a 100% improvement in performance, seems like it should be worthy of a release note, no?

**Comment by wojtek-t** [comment]
> @wojtek-t if this is really a 100% improvement in performance, seems like it should be worthy of a release note, no?

I don't agree. Release notes are for users and/or cluster admins. This change doesn't impact either of those in any visible way.

**Pull request merged**